### PR TITLE
LibJS: Implement basic scoping.

### PR DESCRIPTION
**Pull request by chyza**
Values can now be created and searched for in a ScopeFrame. Searching
will look through all parent ScopeFrames, unless the frame was created
without a parent which is useful for things like function calls where
we don't want the callee to have access to the caller's scope.
Some things are still wrong, functions should have access to the scope
they're defined in for example, but let's not get carried away.

### CODE REVIEW

**Comment by awesomekling** [comment]
Please observe https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md

You'll need to reindent this, the easiest is to just use `clang-format`.

File: Libraries/LibJS/Value.h [hunk 1]
```diff
@@ -36,6 +36,7 @@ namespace JS {
 class Value {
 public:
     enum class Type {
+        NotDefined,
```
Comment (awesomekling) [inline Libraries/LibJS/Value.h]: This class is meant to represent a JavaScript value, and there is no such thing as a "not defined" value in JS.
If you want to communicate the absence of a value somewhere, please use a wrapper like `Optional<Value>`

File: Libraries/LibJS/Interpreter.h [hunk 2]
```diff
@@ -27,33 +27,45 @@
...
     void do_return();
 
 private:
+    void enter_scope_unparented(const ScopeNode&);
     void enter_scope(const ScopeNode&);
     void exit_scope(const ScopeNode&);
 
     Vector<ScopeFrame> m_scope_stack;
-
-    Object* m_global_object { nullptr };
```
Comment (awesomekling) [inline Libraries/LibJS/Interpreter.h]: Please do not remove the global object.

File: Libraries/LibJS/Interpreter.h [hunk 3]
```diff
@@ -27,33 +27,45 @@
...
 class Interpreter {
 public:
     Interpreter();
     ~Interpreter();
 
-    Value run(const ScopeNode&);
+    Value run(const ScopeNode& scope_node, bool parented = true);
 
-    Object& global_object() { return *m_global_object; }
-    const Object& global_object() const { return *m_global_object; }
```
Comment (awesomekling) [inline Libraries/LibJS/Interpreter.h]: Please do not remove the global object.

File: Libraries/LibJS/Interpreter.h [hunk 4]
```diff
@@ -27,33 +27,45 @@
...
-struct ScopeFrame {
-    const ScopeNode& scope_node;
+class ScopeFrame {
+public:
+    ScopeFrame(const ScopeNode& scope_node, const ScopeFrame* parent);
+
+    Value get_var(String name) const;
+    void put_var(String name, Value value);
+
+    const ScopeNode& get_scope_node() const { return m_scope_node; }
```
Comment (awesomekling) [inline Libraries/LibJS/Interpreter.h]: Please do not use `get_` prefixes, per https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#wrong-3

File: Libraries/LibJS/Interpreter.cpp [hunk 5]
```diff
@@ -28,21 +28,49 @@
...
+    if (!found.exists()) {
+        found = m_vars.get(name).value_or(js_not_defined());
+    }
+
+    return found;
+}
+
+void ScopeFrame::put_var(String name, Value value)
+{
+    m_vars.set(name, value);
```
Comment (awesomekling) [inline Libraries/LibJS/Interpreter.cpp]:
```suggestion
    m_vars.set(move(name), move(value));
```

**Comment by chyza** [comment]
Applied the requested changes, took a while because i messed up a git command and nearly nuked my repo but it should all be proper now.

**Comment by Sobak** [comment]
Aside from conflicts that arose in the meanwhile please take a look at the diff of your change. I don't know if this is because of different clang-format settings but your branch has a bunch of unnecessary changes which make diff longer and harder to follow, including the files which were not modified otherwise.

Also, I believe that as per convention commit subject should not end with a period.

**Comment by chyza** [comment]
The amount of things that changed with the garbage collector is enough to completely change the implementation of this (the current one with conflicts resolved compiles and runs but won't really \*work\*) so i don't see the point of leaving this specific pull request open. But hey, at least i had fun writing this :)

**Comment by Sobak** [comment]
@chyza I hope that my comment didn't sound rude. I'm just a bit sensitive when it comes to a good diffs which focus on the actual change and I can clearly see that Andreas tends to do the same. It's cool that you at least considered it fun to work on!

Unfortunately writing C++ it's totally outside of my skills so all I can do is to follow the project's development closely or nitpick some tiny details like git diffs :D

**Pull request closed without merging**